### ISSUE TITLE
Auto merge not processing PRs

### ISSUE THREAD

**Issue by ondrejstasek** [bobvanderlinden/probot-auto-merge#708]
Hi.
We have probot-auto-merge installed inside our organization. Assigned to 2 repos. For one repo it works fine, we had no problems setting it up. Just work out-of-the-box. But for the second repo we are unable to get it running. In both repos we have the same `.github/auto-merge.yml`.
In the first repo, `probot-auto-merge` runs without a problem, reports back the status as a check.
In the second repo, it just don't do anything. No check. Seems like it does not process the PRs at all.
Is there a way how to debug whats happening? Any help or advise will be very appreciated.

**Comment by tamj0rd2** [comment]
I've also started having this issue within the last few days. I've enabled the check to try to debug what's going on, but the check doesn't seem to work either. Here's my [auto-merge.yml](https://github.com/tamj0rd2/ncdc/blob/master/.github/auto-merge.yml)

**Comment by terkin** [comment]
started having this issue today, everything worked fine before, I havent changed yml file, bot just stopped pick up PR with label..

**Comment by tamj0rd2** [comment]
This seems to be working again for me now

**Comment by bobvanderlinden** [comment]
I saw that there were a lot of request time outs toward GitHub. I restarted the application so that GitHub wasn't being bombarded with requests for a short while. I still saw request timeouts after that. Not sure what has helped.

It could be a problem at GitHub, but GitHub Status doesn't show much downtime. It could also be hitting some restriction of Heroku, but I have not received a notification about this. It could also be a bug in some way.

When this ever happens again, please also check GitHub status: https://www.githubstatus.com/

**Comment by kielfriedt** [comment]
Any updates? we have many repos using the merge bot but some work without issue other dont have the check at all. Each repo has the same yml file. How do we debug this?

**Comment by sudarakas** [comment]
the same issue, but worked fine in yesterday but now not working

**Comment by nhojpatrick** [comment]
any issues today?

github status showing okay...

trying to setup and configure on a few projects but not luck today

**Comment by rogerluan** [comment]
I'm experiencing this as well. 
@bobvanderlinden could this be a problem with Heroku server being unreliable?

**Comment by bobvanderlinden** [comment]
Yes, it is still running on a free instance of heroku. Everything currently happens in-memory. It is hard for me to see whether this is actually the problem. Suggestions are welcome! If you really rely on the functionality, is suggest hosting your own instance. Or sponsor hosting of a higher tier heroku instance.

**Comment by rogerluan** [comment]
@bobvanderlinden I've been spinning up several PRs since yesterday when I started testing this, and this may have swamped your instance with requests even more than usual, which may be the reason behind my increasing number of down times (the bot not picking up PRs).

I'll try to play around with this when I have time, fork & deploy to my own server instance and see if things get better :) 

Thanks for the input! I'll report back, letting you know how this goes.

**Comment by bobvanderlinden** [comment]
Hmm, auto-merge does maintain a queue per repository. Maybe best would be to maintain a queue per user/organisation. In the logs I found in the past that there are quite a few organisations that enable auto-merge for all of their private repositories. Auto-merge gets all events of those repositories, needs to fetch configuration without there being a configuration and conclude that nothing should be done.

1. This is something that can be optimized quite a bit. It probably needs a queue per user/organisation and a queue per repository within that user/organisation queue.

2. In addition, auto-merge could cache in-memory whether a repository doesn't have a configuration. It's a bit tricky, as newly set up repositories will be incorrectly marked as not having a configuration until the cache-time runs out.

3. That said, I just checked and "request timeout"s are not entirely solved. I'm not sure yet whether this is due to limitations of Heroku, some kind of throttling from GitHubs end or something else.

1 helps a bit. 2 can help quite a lot, but probably not without degrading UX. 3 is still quite unclear, but may be resolved by upgrading Heroku to a pro account.

**Comment by rogerluan** [comment]
Could the time outs be related to rate limiting? https://developer.github.com/v3/#rate-limiting

Maybe GitHub is blocking the stream of requests coming from the same IP?

I don't think my specific problem is related to those work queues, because I'm only testing this in 1 (private) repo, so whether it has a queue per repo, or per org, it wouldn't matter 🤔

**Comment by rogerluan** [comment]
@bobvanderlinden so I've been testing this bot in my own self-hosted heroku instance and I noticed that every now and then the bot stops picking up PRs, even if it was picking up the same PRs correctly earlier (for example when the PR wasn't ready to be merged, it picked it up, but now it's not triggering checks anymore).

As I write, the current state is:

### PR 1

I change labels on it and it should be ready to be merged, although the bot doesn't process the PR. By checking out the `Advanced` tab in the GitHub Apps page, and by using `heroku logs --tail`, I see that GitHub is doing fine: the webhook is triggered, and the bot responds with a Status 200 OK, as it can be seen below:

![image](https://user-images.githubusercontent.com/8419048/90415970-3ce47280-e088-11ea-8c02-eefc5be5821b.png)

However, nothing happens below the last line shown in the screenshot (ignore what is partially visible there, it's unrelated to the request).

### PR 2

I change labels on it and it should be ready to be merged except it doesn't have latest master (and it's configured to update the branch). Exact same scenario as PR 1, it's not picking up the PR, not merging latest master, etc.

### PR 3

This one is not ready to be merged, but as I update it, the bot picks it up correctly and updates the check accordingly (showing the conditions that it doesn't meet), example (tail only):

<img width="1920" alt="image" src="https://user-images.githubusercontent.com/8419048/90416315-b2e8d980-e088-11ea-90aa-94fc6714ffd3.png">

I _think_ I haven't witnessed a PR that was "stale" and started being picked up by the bot. So I can't tell if they can recover from such state.
It's weird cuz at the same time that there are PRs that are working just fine, others aren't being picked up.

Is there some fancy queueing happening to process work loads that could have some bug in it? Can you think of semaphores, deadlocks, or something that could be causing certain PRs to remain stale?

**Comment by bobvanderlinden** [comment]
> Does this mean that if there are say 3 PRs that are ready to be merged but need master merged into them, the bot would merge master into 1 of them only, and once it is merged, then merge the 2nd, etc? 😲😲

Yes that is exactly right. This is one of the goals with auto-merge, as doing this without such mechanisms will result in a huge amount of overhead (unnecessary branch-updates and thus unnecessary CI running). For 5 open PRs that would potentially mean 15 merge commits (5+4+3+2+1).

I do have to say that this mechanism is the more 'flaky' feature of auto-merge, as everything happens in-memory. I want to move it over to something like Redis, but I do not want to bear the financial costs of having a Redis instance running in addition to probot-auto-merge. For most applications, in-memory is fine, but all queues will be forgotten whenever auto-merge is updated (thus restarting). This is why I thought most of these issues are due to hosting on Heroku.

> Now looking again at the screenshot 1 (PR 1), I see the queued action, however that status wasn't reported in the PR checks (should it?).

It cannot report on the PR pending, because for that it first needs to read the configuration. Reading the configuration should also be queued (along with any other information fetched from GitHub), otherwise events from one repository can still starve other repositories.

> Usually when the PR has pending checks, auto-merge will update its PR status to "pending" (or "waiting", can't remember the exact wording), but this wasn't the case 🤔

Yes, it does that if it is waiting on that particular PR to be out of pending state. All other queued PRs will not have their status checks updated. This is by design (in terms of performance). Updating status checks of all other PRs is quite heavy as GitHub doesn't support updating multiple checks at once. (it requires querying a PR and then updating the check for the HEAD commit, for each of the other PRs).



That all said, I now saw this issue happening again and noticed Heroku reporting on request timeouts. The requests that GitHub executes toward auto-merge took more than 30s. I can imagine this happening, as handing an event happens within the request. If auto-merge needs to wait for an event to be processed, then it will leave the request open during that time. With enough traffic and with the addition of needing to wait for pending checks, I can imagine this being one reason events are missed by auto-merge. auto-merge needs to accept all events, add them to a queue immediately and solely handle the queue tasks. Letting GitHub know whether an event was handled or not is of no value.

**Comment by rogerluan** [comment]
>Yes that is exactly right. 

This is amazing! 🎉 
Your design decisions all make sense, and now I understand the queueing system better.

With your last paragraph, are you saying you spotted the issue? 🙈  I got a bit confused in the last sentence, but what I understood is that if the body is busy waiting for its queues, it won't be available to "reply" to github webhook events, thus dropping those events. Is that what's happening?

**Comment by bobvanderlinden** [comment]
I spotted something that doesn't seem right. Not sure how much it will help. I can imagine that GitHub has too many requests open toward auto-merge and thus will not send any more. I'm not 100% what the consequence is of Heroku cutting off long-running connections, but it's actually a good thing it does that.
So, not sure this fixes the issue, but it is the best lead.
It also allows me to do improve the queuing some more. At the moment each repository has a queue, but some organisations have a large number of repositories, taking away time from other repositories: some organisations enable auto-merge for all repos without adding auto-merge configuration.
I'd like to have a queue of organisations and have each organisation-queue have a queue of repositories and each repository-queue have a queue of pull requests.

> Letting GitHub know whether an event was handled or not is of no value.

My intention was to respond with 200 whenever an event was handled correctly and something else when it was not. However, this response code is of no value. It should just respond with 200 if auto-merge has queued the event successfully.

**Comment by rogerluan** [comment]
Skipping repositories that don't have `auto-merge.yml` sounds like something we should do imo, to avoid needlessly processing all PRs of an organization (possibly dozens, or hundreds) 😱 

>It should just respond with 200 if auto-merge has queued the event successfully.

Makes sense! 👍

**Comment by rogerluan** [comment]
It seems like PRs aren't being picked up again 😬 like at all... new installation in new organization in new repository. I've reviewed 4 times to be sure I set everything up correctly and I think I did 🤷‍♂️
